// wiremock-captain/dist/types/externalTypes.js
"use strict";
// Copyright (c) WarnerMedia Direct, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.WireMockFault = exports.MatchingAttributes = exports.EndpointFeature = exports.DelayType = exports.BodyType = void 0;
var BodyType;
(function (BodyType) {
    BodyType["Default"] = "jsonBody";
    BodyType["Body"] = "body";
    BodyType["Base64Body"] = "base64Body";
})(BodyType = exports.BodyType || (exports.BodyType = {}));
var DelayType;
(function (DelayType) {
    DelayType["CHUNKED_DRIBBLE"] = "CHUNKED_DRIBBLE";
    DelayType["FIXED"] = "FIXED";
    DelayType["LOG_NORMAL"] = "LOG_NORMAL";
    DelayType["UNIFORM"] = "UNIFORM";
})(DelayType = exports.DelayType || (exports.DelayType = {}));
var EndpointFeature;
(function (EndpointFeature) {
    EndpointFeature["Default"] = "url";
    EndpointFeature["UrlPath"] = "urlPath";
    EndpointFeature["UrlPathPattern"] = "urlPathPattern";
    EndpointFeature["UrlPattern"] = "urlPattern";
})(EndpointFeature = exports.EndpointFeature || (exports.EndpointFeature = {}));
var MatchingAttributes;
(function (MatchingAttributes) {
    MatchingAttributes["BinaryEqualTo"] = "binaryEqualTo";
    MatchingAttributes["Contains"] = "contains";
    MatchingAttributes["DoesNotMatch"] = "doesNotMatch";
    MatchingAttributes["EqualTo"] = "equalTo";
    MatchingAttributes["EqualToJson"] = "equalToJson";
    MatchingAttributes["EqualToXml"] = "equalToXml";
    MatchingAttributes["Matches"] = "matches";
    MatchingAttributes["MatchesJsonPath"] = "matchesJsonPath";
})(MatchingAttributes = exports.MatchingAttributes || (exports.MatchingAttributes = {}));
var WireMockFault;
(function (WireMockFault) {
    WireMockFault["CONNECTION_RESET_BY_PEER"] = "CONNECTION_RESET_BY_PEER";
    WireMockFault["EMPTY_RESPONSE"] = "EMPTY_RESPONSE";
    WireMockFault["MALFORMED_RESPONSE_CHUNK"] = "MALFORMED_RESPONSE_CHUNK";
    WireMockFault["RANDOM_DATA_THEN_CLOSE"] = "RANDOM_DATA_THEN_CLOSE";
})(WireMockFault = exports.WireMockFault || (exports.WireMockFault = {}));
