// wiremock-captain/dist/ResponseModel.js
"use strict";
// Copyright (c) WarnerMedia Direct, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWireMockResponse = void 0;
const externalTypes_1 = require("./types/externalTypes");
function createWireMockResponse(response, features) {
    const { body, headers, status } = response;
    const mockedResponse = {
        status,
    };
    const bodyType = features?.responseBodyType ?? externalTypes_1.BodyType.Default;
    if (body) {
        mockedResponse[bodyType] = body;
    }
    mockedResponse.headers = {
        ...(bodyType === externalTypes_1.BodyType.Default && {
            'Content-Type': 'application/json; charset=utf-8',
        }),
        ...headers,
    };
    const delay = features?.responseDelay;
    switch (delay?.type) {
        case externalTypes_1.DelayType.CHUNKED_DRIBBLE:
            mockedResponse.chunkedDribbleDelay = {
                numberOfChunks: delay.numberOfChunks,
                totalDuration: delay.totalDuration,
            };
            break;
        case externalTypes_1.DelayType.FIXED:
            mockedResponse.fixedDelayMilliseconds = delay.constantDelay;
            break;
        case externalTypes_1.DelayType.LOG_NORMAL:
            mockedResponse.delayDistribution = {
                type: 'lognormal',
                median: delay.median,
                sigma: delay.sigma,
            };
            break;
        case externalTypes_1.DelayType.UNIFORM:
            mockedResponse.delayDistribution = {
                type: 'uniform',
                lower: delay.lower,
                upper: delay.upper,
            };
            break;
        default:
    }
    return mockedResponse;
}
exports.createWireMockResponse = createWireMockResponse;
