// wiremock-captain/dist/WireMock.js
"use strict";
// Copyright (c) WarnerMedia Direct, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.WireMock = void 0;
const axios_1 = require("axios");
const RequestModel_1 = require("./RequestModel");
const ResponseModel_1 = require("./ResponseModel");
const utils_1 = require("./utils");
// endpoint where wiremock stores mocks
const WIREMOCK_MAPPINGS_URL = '__admin/mappings';
// endpoint that records all the incoming requests
const WIREMOCK_REQUESTS_URL = '__admin/requests';
// endpoint that records all the scenario information
const WIREMOCK_SCENARIO_URL = '__admin/scenarios';
const HEADERS = { 'Content-Type': 'application/json' };
class WireMock {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
    }
    /**
     * Creates a new stub with desired request and response match
     * @param request Request object for the stub mapping
     * @param response AxiosResponse object for the stub mapping
     * @param features Additional options to be used for creation of stub mapping
     * @returns Created wiremock stub mapping. Contains `id` which is needed to delete a mapping
     */
    async register(request, response, features) {
        const mockedRequest = (0, RequestModel_1.createWireMockRequest)(request, features);
        const mockedResponse = (0, ResponseModel_1.createWireMockResponse)(response, features);
        let mock = {
            request: mockedRequest,
            response: mockedResponse,
        };
        if (features?.stubPriority) {
            mock.priority = features.stubPriority;
        }
        if (features?.scenario) {
            mock = { ...mock, ...features.scenario };
        }
        if (features?.webhook) {
            mock.postServeActions = [
                {
                    name: 'webhook',
                    parameters: {
                        method: features.webhook.method,
                        url: features.webhook.url,
                        ...(features.webhook.headers && { headers: features.webhook.headers }),
                        ...(features.webhook.body && {
                            body: (0, utils_1.getWebhookBody)(features.webhook.body),
                        }),
                        ...(features.webhook.delay && {
                            delay: (0, utils_1.getWebhookDelayBody)(features.webhook.delay),
                        }),
                    },
                },
            ];
        }
        if (features?.fault) {
            mock.response = { fault: features.fault };
        }
        const wiremockResponse = await axios_1.default.post(this.makeUrl(WIREMOCK_MAPPINGS_URL), mock, {
            headers: HEADERS,
        });
        return wiremockResponse.data;
    }
    /**
     * Removes all the existing stubs and logs of incoming requests
     */
    clearAll() {
        return Promise.all([this.clearAllMappings(), this.clearAllRequests()]);
    }
    /**
     * Removes all the non-default mappings (not defined in backing store)
     * and logs of incoming requests
     */
    clearAllExceptDefault() {
        return Promise.all([this.resetMappings(), this.clearAllRequests()]);
    }
    /**
     * Removes all existing stubs
     */
    async clearAllMappings() {
        return await axios_1.default.delete(this.makeUrl(WIREMOCK_MAPPINGS_URL));
    }
    /**
     * Removes log of all past incoming requests
     */
    async clearAllRequests() {
        return await axios_1.default.delete(this.makeUrl(WIREMOCK_REQUESTS_URL));
    }
    /**
     * Deletes a stub mapping
     * @param id ID of the stub to be deleted. Can be obtained when from response of `register()`
     */
    async deleteMapping(id) {
        return await axios_1.default.delete(this.makeUrl(WIREMOCK_MAPPINGS_URL + '/' + id));
    }
    /**
     * Returns all the request and response mappings attached to the wiremock instance
     * @returns Collection of all mappings for the wiremock instance
     */
    async getAllMappings() {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_MAPPINGS_URL));
        const responseJson = response.data;
        return responseJson.mappings;
    }
    /**
     * @returns List of all requests made to the mocked instance
     */
    async getAllRequests() {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_REQUESTS_URL));
        const responseJson = response.data;
        return responseJson.requests;
    }
    /**
     * @returns List of all scenarios in place for the mocked instance
     */
    async getAllScenarios() {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_SCENARIO_URL));
        const responseJson = response.data;
        return responseJson.scenarios;
    }
    /**
     * Returns information about the mocked request and response corresponding to the `id`
     * @param id Mapping ID to get the mapping info for
     * @returns Single object mapping corresponding to the input `id`
     */
    async getMapping(id) {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_MAPPINGS_URL + '/' + id));
        return await response.data;
    }
    /**
     * Returns list of request(s) made to the WireMock API
     * @param method Method to match the request(s) made against
     * @param endpointUrl URL to get the request(s) made against
     * @returns List of wiremock requests made to the endpoint with given method
     */
    async getRequestsForAPI(method, endpointUrl) {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_REQUESTS_URL));
        const body = response.data;
        return (body.requests
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            .filter((r) => (0, utils_1.filterRequest)(method, endpointUrl, r)));
    }
    /**
     * Returns list of unmatched request(s)
     * @returns List of wiremock requests made that did not match any mapping
     */
    async getUnmatchedRequests() {
        const response = await axios_1.default.get(this.makeUrl(WIREMOCK_REQUESTS_URL + '/unmatched'));
        const body = response.data;
        return body.requests;
    }
    /**
     * Resets all the scenarios to the original state
     */
    async resetAllScenarios() {
        return await axios_1.default.post(this.makeUrl(WIREMOCK_SCENARIO_URL + '/reset'));
    }
    /**
     * Restores stub mappings to the defaults defined back in the backing store
     */
    async resetMappings() {
        return await axios_1.default.post(this.makeUrl(WIREMOCK_MAPPINGS_URL + '/reset'));
    }
    makeUrl(endpoint) {
        return new URL(endpoint, this.baseUrl).href;
    }
}
exports.WireMock = WireMock;
